### ISSUE TITLE
Understanding how training in batch occurs

### ISSUE THREAD

**Issue by kushalj001** [bentrevett/pytorch-seq2seq#94]
Hello @bentrevett
I am trying to understand how batch processing works with respect to training neural networks. I had some doubt and it would be great if you could look into this.
This is the dataset from all your tutorials.
![image](https://user-images.githubusercontent.com/32245327/78457800-7b295880-76ca-11ea-9e7b-a4578040adf5.png)
The target shape is 
![image](https://user-images.githubusercontent.com/32245327/78457836-c479a800-76ca-11ea-9388-5ceb1d426560.png)
Following is how the target tensor looks
![image](https://user-images.githubusercontent.com/32245327/78457844-e2470d00-76ca-11ea-8723-13a4fcf7f08d.png)
The tensor of targets looks like this with index 2 being the sos token, 3 being the eos and 1 is the pad. The training examples are arranged in columns so to say.
This is the target sequence for the first training example
![image](https://user-images.githubusercontent.com/32245327/78457902-410c8680-76cb-11ea-8090-1d0dfec7a113.png)
While predicting our output, we stack our predictions sequentially in the for loop
however, when we compute the loss, we do the following
![image](https://user-images.githubusercontent.com/32245327/78457920-67322680-76cb-11ea-82df-947dfc9fe25b.png)
 Comparing the tensor values with x.trg above, we see that this vector is not arranged in the sequence of the sentence, as shown above, rather the values at index 1 for all the sentences in the target are together, followed by values at index 2 and so on.

When we get the output from our model, are the outputs stacked in a similar way? Or are they arranged sequentially for all the targets; like the output of training example followed by another example in the sequence of actual sentences? 
Thank you

**Comment by bentrevett** [comment]
Yes, the outputs are arranged in the same way, e.g. if you had a batch size of 32 and a sequence length of 20 with a vocab size of 100 then they would be arranged so that the first 32 elements would be the 100-dimensional predictions for the first token of each of the 32 sequences, the next 32 would be the 100-dimensional predictions for the second token of each of the 32 sequences, etc.

This is just how PyTorch handles reshaping things with `.view` and if we somehow reshaped them so the first 20 elements were the first sequence, the next 20 were the second sequence, etc. then you would get the same loss value.

**Comment by kushalj001** [comment]
Thanks @bentrevett 
Also, I want to keep track of a metric other than the loss during training. If I calculate the bleu score for the predictions and the target by passing them to the torchtext method, it would work, right? As in the alignment would still be correct as in case of the loss mentioned above, right? 
Thanks a lot.

**Comment by bentrevett** [comment]
BLEU is a bit more difficult. You would have to reshape the predictions to `[batch, sequence, vocab]` and then take the argmax across the final dimension. Then reshape the targets to `[batch, sequence]` and then remove all padding from the predictions and targets. It is relatively slow to calculate compared to the loss.
